### PR TITLE
core: resolve circular dependency

### PR DESCRIPTION
**Pull request by maribu**
### Contribution description

Dependency graph before:

```
     +-------------------------------------+
     |                                     |
     |                                     V
+----------+        +-------+         +---------+
| thread.h | -----> | msg.h | ------> | sched.h |
+----------+        +-------+         +---------+
     ^                                     |
     |                                     |
     +-------------------------------------+
```

Dependency graph after:

```
                 +--------------+
                 | core_types.h |
                 +--------------+
                    ^   ^   ^
                    |   |   |
      +-------------+   |   +--------------+
      |                 |                  |
+----------+            |             +---------+
| thread.h |            |             | sched.h |
+----------+            |             +---------+
                        |                  ^
                    +-------+              |
                    | msg.h |--------------+
                    +-------+
```

### Testing procedure

This PR doesn't change code. It only relocates it. Hence, binaries shouldn't change.

### Issues/PRs references

This circular dependency became an issue in https://github.com/RIOT-OS/RIOT/pull/16174

### CODE REVIEW

**Review by benpicco — APPROVED** [review]
Nice cleanup!

**Review by kaspar030 — CHANGES_REQUESTED** [review]
Pleas don't merge too quick, I'd like to take a closer look.

**Comment by benpicco** [comment]
Yea Murdock has found some isssues.

**Comment by kaspar030** [comment]
I'm not in favour of shuffling the headers around like this.

How about just forward declaring msg_t? The only reason thread.h needs msg_t is a single pointer. -> https://github.com/RIOT-OS/RIOT/pull/16477

**Comment by aabadie** [comment]
This seems cleaner than #16477 (and needs a rebase).

**Comment by maribu** [comment]
> I'm not in favour of shuffling the headers around like this.
> 
> How about just forward declaring msg_t? The only reason thread.h needs msg_t is a single pointer. -> #16477

That doesn't solve the issue that `sched.h` depends on `thread.h` depends on `sched.h` ...

File: core/include/core_types.h
Comment (aabadie) [inline core/include/core_types.h:77]: You could make this a `@brief` to fix the Doxygen error and be consistent with how other structs are documented in this file:

```suggestion
 * @brief Thread states supported by RIOT
 *
 *       Keep in sync with OpenOCD src/rtos/riot.c
 */
typedef enum {
    STATUS_STOPPED,                 /**< has terminated                           */
    STATUS_ZOMBIE,                  /**< has terminated & keeps thread's thread_t */
    STATUS_SLEEPING,                /**< sleeping                                 */
    STATUS_MUTEX_BLOCKED,           /**< waiting for a locked mutex               */
    STATUS_RECEIVE_BLOCKED,         /**< waiting for a message                    */
    STATUS_SEND_BLOCKED,            /**< waiting for message to be delivered      */
    STATUS_REPLY_BLOCKED,           /**< waiting for a message response           */
    STATUS_FLAG_BLOCKED_ANY,        /**< waiting for any flag from flag_mask      */
    STATUS_FLAG_BLOCKED_ALL,        /**< waiting for all flags in flag_mask       */
    STATUS_MBOX_BLOCKED,            /**< waiting for get/put on mbox              */
    STATUS_COND_BLOCKED,            /**< waiting for a condition variable         */
    STATUS_RUNNING,                 /**< currently running                        */
    STATUS_PENDING,                 /**< waiting to be scheduled to run           */
    STATUS_NUMOF                    /**< number of supported thread states        */
} thread_status_t;
```

**Comment by maribu** [comment]
@chrysn For this to be compatible with rust, `riot_sys` should use `thread_t` instead of `struct _thread`, right? That should also work with both current `master` and this PR.

**Comment by chrysn** [comment]
Rust has no difference between type names and named structs, and it looks at all the headers at once, so I don't think it makes any difference. Bindings already look like this:

```rust
#[repr(C)]
pub struct _thread {
    pub sp: *mut c_char,
    pub status: thread_status_t,
    // ...
}
pub type thread_t = _thread;
```

where `thread_t` is purely an alias for `_thread` and usable everywhere interchangeabl. That is, functions may declare either in their signature, but a user would use `thread_t` everywhere and it would be accepted -- it could just as well say `pub struct thread_t { ... }` / `pub type _thread = thread_t;`.

**Comment by maribu** [comment]
The issue is that with this PR, `struct _thread` is dropped and only `thread_t` is provided. It is a pity that Rust internally references the internal alias `_thread`. Would it be possible to drop this, so that this rust example would still build with this PR?

**Comment by chrysn** [comment]
The Rust bindings (of riot-sys) are always automatically regenerated
from the used RIOT code. When `struct _thread` goes away, there will not
be a `_thread` any more in riot-sys.

Aaah I see -- riot-wrappers uses `_thread`. That's unfortunate, it
shouldn't have been doing this. (In general, riot-sys contains a lot of
symbols that one should rather not use -- basically everything there is
no RIOT documentation for. That's the nature of going from an
unnamespaced language to a namespaced one).

Fix coming, sorry for both the bug and me not getting it originally.

**Comment by chrysn** [comment]
Please rebase onto (I prefer merging in, but was told GitHub doesn't
handle this workflow) https://github.com/RIOT-OS/RIOT/pull/17514 once
that's merged. (Let's hope there are a few fast ones in the CI queue).

**Comment by miri64** [comment]
Huh, what happened here? Seems somewhat important!

**Comment by maribu** [comment]
I still think this should go in

**Comment by maribu** [comment]
Rebased and reordered the commits so that the addition of missing `#include`s (which previously were pulled in indirectly) are added first, and the actual change comes afterwards. That way bisecting should be easier.

**Comment by Teufelchen1** [comment]
Did we have another issue with circular dependency since #16174?
I feel a bit like not having enough insights into `/core/` to have an opinion but if taken with a grain of salt, this looks like a reasonable change. I would approve it :> 
Maybe you can rebase if you are still interested maribu?

**Comment by maribu** [comment]
We eventually have worked around the issue. But generally, circular dependencies in C headers are a bit messy, so this may be useful later down the road.

Maybe close this for now and pick this up again when someone bumps their feet on this again?

**Comment by Teufelchen1** [comment]
Closed.
- Problem has been identified
- Multiple proposed solution, including this, one didn't get "enough approval"
- Solution is not urgently needed
- Will be revisited if someone finds another circular dependency error / issue

Thanks for your proposal @maribu! :)

**Comment by Teufelchen1** [comment]
Closed.
- Problem has been identified
- Multiple proposed solution, including this, one didn't get "enough approval"
- Solution is not urgently needed
- Will be revisited if someone finds another circular dependency error / issue

Thanks for your proposal @maribu! :)

**Pull request closed without merging**
